### PR TITLE
Refactor folder sync

### PR DESCRIPTION
**Pull request by ChristophWurst**
~~Partially~~ mostly implements #318
Fixes https://github.com/nextcloud/mail/issues/244, #318, https://github.com/nextcloud/mail/issues/335, https://github.com/nextcloud/mail/issues/290

- Sync
  - [x] Use horde sync to get new/changed/vanished messages
  - [x] Implement message mapper (reuse old code if possible)
  - [x] Implement sync mechanism in the client
  - [x] Implement sync for the unified inbox
  - [x] Update unified inbox if individual ones change
  - [x] Implement sync for search mailboxes – should work now *fingers crossed*
  - [x] Implement pagination for search mailboxes
  - [x] test search
- Pagination
  - [x] Implement date-based cursor-based pagination on the server
  - [x] Implement new pagination on the client
  - [x] Implement pagination for the unified inbox

- Necessary follow-ups
  - Kill background sync. It's extremely buggy in its current form. If the new horde sync approach is fast enough, I'm planning to just let the current folder view refresh its content once a minute. If the folder is a specific inbox, we refresh the unified one. This should cover everything.

### CODE REVIEW

**Comment by ChristophWurst** [comment]
This is gonna be a big one. I'm already sorry for that, reviewers. I've tried to split it up but this I'm afraid that would be 5x more work.

**Comment by ChristophWurst** [comment]
Wow, my IMAP server doesn't even provide the capabilities that make horde sync fast (MODSEQ for example) and still it is noticeably faster 😮 🚀

**Comment by ChristophWurst** [comment]
Unified inbox synchronization and pagination seems to work. Still, I have to tweak it a bit to be as efficient as I desired it (remove some redundant requests). Looks promising so far 🚀

**Comment by ChristophWurst** [comment]
For those who want to test it on their production instance, here's a signed archive: [mail.tar.gz](https://github.com/nextcloud/mail/files/1035826/mail.tar.gz)

**Comment by Gomez** [comment]
Notice: Nextcloud 12 is needed, is it a hard requirement?

**Comment by ChristophWurst** [comment]
> Notice: Nextcloud 12 is needed, is it a hard requirement?

Yes, because the next version of Mail will already support NC13. That covers what we agreed on a few months back: support the latest development and released version.

For this PR it's probably not critical to run NC11, but stuff like #3 will fail hard on NC11 and lower.

**Comment by ChristophWurst** [comment]
I've just found out that the ajax call to sync a mailbox does not always send all required parameters, hence the request results in a HTTP 400 sometimes. Debugging this at the moment …

**Comment by ChristophWurst** [comment]
Should be fixed with the latest commit 
[mail.tar.gz](https://github.com/nextcloud/mail/files/1035975/mail.tar.gz)

**Comment by ChristophWurst** [comment]
- [ ] BUG: sometimes messages cannot be deleted

**Comment by ChristophWurst** [comment]
@Gomez @jancborchardt did anybody of you have the time to test this PR? I've used it for two weeks now and the only annoying bug I could find was that I was unable to delete messages in my inbox as soon as I synced it once. This should be fixed now however.

Let me know if there are any other bugs/regressions :)

**Comment by jancborchardt** [comment]
Ok, let's test this for a bit then. :) cc @lcalaresu if you want to join

**Comment by lcalaresu** [comment]
> Ok, let's test this for a bit then. :) cc @lcalaresu if you want to join

It's installed on my production server... Will definitely come back here if there is a problem 😉

**Comment by lcalaresu** [comment]
Please forgive my ignorance.
I am using it in my production server, no problem so far but... is there any specific things I could do to "see" the difference ?

**Comment by ChristophWurst** [comment]
> Please forgive my ignorance.
> I am using it in my production server, no problem so far but... is there any specific things I could do to "see" the difference ?

Depending on how many account you're using, synchronizing (aka checking for new mails) should be faster as this parallelizes a lot. Same applies to the initial page load and the pagination in the message list. Functionality-wise it should be more or less the same as before :)

**Comment by lcalaresu** [comment]
Sorry I spoke too fast... I ran today into some problems.

1. The first one is more of a disturbance. 
Everytime I click on a message, there is some weird icons appearing in unexpected places while loading and disappearing when the message is loaded. Not sure I have seen that before:
dots  stars  wastebasket

2. I also had a problem with a message that was stuck as "unseen". And got the JS error:
unseen.
I reloaded the whole page, but at that time the message completely disappeared (it is still on the server though). It made me realised **some** messages were missing on my mailbox but still on the server :sob: 

3. My server logs are filled with the messages:

> Error	PHP	Undefined index: messages at /apps2/mail/lib/Folder.php#172	2017-06-14 12:14:46
Error	PHP	Undefined index: messages at /apps2/mail/lib/Folder.php#171	2017-06-14 12:14:46
Error	PHP	Undefined index: unseen at /apps2/mail/lib/Folder.php#170	2017-06-14 12:14:46
Error	PHP	Undefined index: messages at /apps2/mail/lib/Folder.php#172	2017-06-14 12:14:46
Error	PHP	Undefined index: messages at /apps2/mail/lib/Folder.php#171	2017-06-14 12:14:46
Error	PHP	Undefined index: unseen at /apps2/mail/lib/Folder.php#170	2017-06-14 12:14:46

**Comment by ChristophWurst** [comment]
Ouch. This obviously needs more testing 😁

**Comment by lcalaresu** [comment]
> Ouch. This obviously needs more testing :grin:

Forget the error (3) in my previous message, it also happens on master so...

**Comment by ChristophWurst** [comment]
@lcalaresu could you please provide steps to reproduce your issues. Apparently I don't see them here on my dev env. If you've found a js error, it would be great to have a stack trace with the original source code and not the minified mail.min.js. You can switch to source files by enabling debug mode in your Nextcloud's config.php. Thanks!

**Comment by lcalaresu** [comment]
@ChristophWurst 
>If you've found a js error, it would be great to have a stack trace with the original source code and not the minified mail.min.js. You can switch to source files by enabling debug mode in your Nextcloud's config.php

Yeah, I forgot to do it when I sent the previous message, but it is activated now.

To reproduce the first problem, I have to use an external mail server (a local one is to fast to actually see the icons changing). 
I just have to select an email that has not been opened yet (so, it is slow enough to see the icons again), and **move the mouse inside the email list while it's loading** (dunno if my explanation is clear enough).

For the second problem, I will check later if I can reproduce it.

**Comment by ChristophWurst** [comment]
I've fixed a bug that caused some weird behavior with anything that has to do with message flags, like marking a message as read/unread/favs and the like. Maybe that resolved your problem too. I cannot reproduce the problem with the provided instructions. But again, thanks a lot for testing! Let me know if the issues still occurs and if possible, please take a short video of that icon flickering. I'm a bit clueless where that could come from.

**Comment by ChristophWurst** [comment]
Fixed another bug that prevented messages being flagged as read when you go through the list with the cursor keys.

**Comment by ChristophWurst** [comment]
Anybody who's used this in production already should try the latest version. It feels a much more stable than previous builds.
Also a note on performance: not all IMAP servers support differential synchronization, like only returning a list of messages that actually have changed. Some, like GMail, always return a list of all messages, as if all have changed. Functionality-wise this is ok, but we're transferring more data than necessary. I do not want to improve this right now (the PR is already bigger than intended) but we can improve that in the future by sending more data to the Nextcloud server and let perform the dirty checking. Assuming that the connection between the client and the nc server is slower than the one from the nc server to the imap server, this should give an additional performance boost.

**Comment by jancborchardt** [comment]
Found 2 issues:

- [ ] in the unified inbox, the colored bars on the left of each message to denote the origin account are missing
- [ ] in any folder, pull to refresh leads to »Error while refreshing messages.«

**Comment by ChristophWurst** [comment]
> in any folder, pull to refresh leads to »Error while refreshing messages.«

Please provide the log entries of nextcloud.com or the browser console. If such an error message is displayed, usually something is logged.

**Comment by ChristophWurst** [comment]
> when changing a folder, there’s no loading feedback. The view stays at the current folder until the new one is loaded and then switches over.

There is, if you switch to a folder that has not been loaded yet. If it has, we also show it but the actual data is already available so it is displayed immediately. This means that the loaded feedback isn't really visible because it's shown only for a few milliseconds, maybe the browser even optimizes that away. There is a tiny lag between switching locally available folders. I think this could be due to the re-rendering of the content views. If there's any other issue with the loaded feedback, please provide a short screen cast, thanks.

**Comment by lcalaresu** [comment]
Lovely! The weird blinking icons do not appear anymore!
I **do not** have the `»Error while refreshing messages.«` message.

I still have two problems:

1. When using a gmail account, I cannot open the `[Gmail]` folder, I get at php error while calling `https://server.com/apps/mail/accounts/4/folders/xUdlmWlsXz%3D%3D/messages?filter=`.  I can give you the stacktrace, but it is huge...

> Horde_Imap_Client_Exception_ServerResponse: The object could not be deleted because it does not exist.


2. On a personal server, I have a JS error every time I change the folder (even base folders like sent, inbox, etc.):

> undefined
Promise rejected (async)
fetchBodies @ foldercontroller.js:155
triggerEvents @ backbone.js:371
triggerApi @ backbone.js:356
eventsApi @ backbone.js:155
Events.trigger @ backbone.js:346
(anonymous) @ foldercontroller.js:79
Promise resolved (async)
loadFolderMessages @ foldercontroller.js:65
(anonymous) @ foldercontroller.js:111
(anonymous) @ underscore.js:768
setTimeout (async)
_.delay @ underscore.js:767
executeBound @ underscore.js:701
bound @ underscore.js:733
showFolder @ foldercontroller.js:110
showFolder @ routecontroller.js:118
triggerEvents @ backbone.js:371
triggerApi @ backbone.js:356
eventsApi @ backbone.js:155
Events.trigger @ backbone.js:346
loadFolder @ folderview.js:90
dispatch @ jquery.min.js:3
r.handle @ jquery.min.js:3

**Comment by ChristophWurst** [comment]
> When using a gmail account, I cannot open the [Gmail] folder, I get at php error while calling https://server.com/apps/mail/accounts/4/folders/xUdlmWlsXz%3D%3D/messages?filter=. I can give you the stacktrace, but it is huge...

Fixed in https://github.com/nextcloud/mail/pull/402 :)

**Comment by ChristophWurst** [comment]
> On a personal server, I have a JS error every time I change the folder (even base folders like sent, inbox, etc.):

Is that a network/server error? Please check your browser's network tab, thx!

**Comment by lcalaresu** [comment]
> Is that a network/server error? Please check your browser's network tab, thx!

On that one, there was no server error as far as I can remember, only the long JS error.
However, I cannot reproduce that problem today... 
Will come back here if it comes back :/

**Comment by ChristophWurst** [comment]
> On that one, there was no server error as far as I can remember, only the long JS error.
> However, I cannot reproduce that problem today...

Fine by me if the bugs disappeared magically :) I couldn't reproduce this either.

**Comment by ChristophWurst** [comment]
- [ ] BUG: deleting messages in the unified inbox doesn't seem to sync changes to the individual ones and thus sometimes deleted messages show up again.

**Comment by ChristophWurst** [comment]
- [ ] BUG: deleting a message select the next one of the the specific folder, not the next in the combined folder.

File: js/controller/foldercontroller.js [hunk 1]
```diff
@@ -152,11 +153,105 @@ define(function(require) {
...
+	 * @param {Message} message
+	 * @param {Folder} currentFolder
+	 * @returns {Promise}
+	 */
+	function deleteMessage(message, currentFolder) {
+		var folders = getSpecificAndUnifiedFolder(message.folder, currentFolder);
+
+		applyOnFolders(folders, function(folder) {
+			// Update total counter and prevent negative values
+			folder.set('total', Math.max(0, folder.get('total')));
```
Comment (ChristophWurst) [inline js/controller/foldercontroller.js]: As a side effect, this fixes https://github.com/nextcloud/mail/issues/335

Comment (ChristophWurst) [inline js/controller/foldercontroller.js]: This will also ensure that the unified/specific unread counters are in sync.

**Comment by ChristophWurst** [comment]
Here's a new build for those who want to test it in production: [mail.tar.gz](https://github.com/nextcloud/mail/files/1119314/mail.tar.gz)

**Comment by ChristophWurst** [comment]
I've been using this in production for about two weeks. Apart from the bugs I listed above (all fixed meanwhile) it seems to be stable and working now. Tested both on the desktop and on my phone. I would love to get some feedback from others too. Hope we can merge this soon to concentrate on stabilization for the next release.

**Comment by jancborchardt** [comment]
I have some issues:

- When sending an email, I get the »Message sent« notification but the reply boxes and button still stays disabled as if it’s still sending.

- The unified inbox sometimes shows different mails than the individual mails, seems to lag behind sometimes (might just be an update artifact though).

- Still get »Error while refreshing messages«

Console log:
```
/nextcloud/apps/mail/js/init.js:25 Starting Mail …
cache.js:47 initializing cache…
http://localhorst/nextcloud/index.php/apps/mail/accounts/2/folders/SU5CT1g%…5B%5D=5108&uids%5B%5D=5102&uids%5B%5D=5094&uids%5B%5D=5077&uids%5B%5D=5050 Failed to load resource: the server responded with a status of 500 (Internal Server Error)
messagesview.js:238 Object
(anonymous) @ messagesview.js:238
http://localhorst/nextcloud/index.php/apps/mail/accounts/1/folders/SU5CT1g%…=21545&uids%5B%5D=21512&uids%5B%5D=21466&uids%5B%5D=21451&uids%5B%5D=21440 Failed to load resource: the server responded with a status of 500 (Internal Server Error)
messagecontroller.js:152 Uncaught (in promise) TypeError: Cannot read property 'account' of undefined
    at Radio.Channel.flagMessage (messagecontroller.js:152)
    at triggerEvents (backbone.js:371)
    at triggerApi (backbone.js:356)
    at eventsApi (backbone.js:155)
    at Radio.Channel.Events.trigger (backbone.js:346)
    at Radio.message.request.then.error (composerview.js:321)
```
And when trying to pull to refresh:
```
jquery.js:8630 GET http://localhorst/nextcloud/index.php/apps/mail/accounts/2/folders/SU5CT1g%…5B%5D=5108&uids%5B%5D=5102&uids%5B%5D=5094&uids%5B%5D=5077&uids%5B%5D=5050 500 (Internal Server Error)
send @ jquery.js:8630
ajax @ jquery.js:8166
syncSingleFolder @ foldersyncservice.js:43
(anonymous) @ foldersyncservice.js:114
syncFolder @ foldersyncservice.js:113
Radio._callHandler @ backbone.radio.js:95
request @ backbone.radio.js:216
_syncMessages @ messagesview.js:236
onScroll @ messagesview.js:170
onWheel @ messagesview.js:183
dispatch @ jquery.js:4435
r.handle @ jquery.js:4121
messagesview.js:238 Object {readyState: 4, responseText: "<!DOCTYPE html>↵<html class="ng-csp" data-placehol…ll your data			</p>↵		</footer>↵	</body>↵</html>↵", status: 500, statusText: "Internal Server Error"}
(anonymous) @ messagesview.js:238
jquery.js:8630 GET http://localhorst/nextcloud/index.php/apps/mail/accounts/1/folders/SU5CT1g%…=21545&uids%5B%5D=21512&uids%5B%5D=21466&uids%5B%5D=21451&uids%5B%5D=21440 500 (Internal Server Error)
```

**Comment by ChristophWurst** [comment]
> jquery.js:8630 GET http://localhorst/nextcloud/index.php/apps/mail/accounts/2/folders/SU5CT1g%…5B%5D=5108&uids%5B%5D=5102&uids%5B%5D=5094&uids%5B%5D=5077&uids%5B%5D=5050 500 (Internal Server Error) send @ jquery.js:8630

Please provide your nextcloud.log entries when that happens. Thanks!

**Comment by ChristophWurst** [comment]
Looks like the errors originate on the server side (all HTTP 500s), hence I'm unable to resolve those problems until I've got error messages and stack traces from your log file.

**Comment by ChristophWurst** [comment]
> When sending an email, I get the »Message sent« notification but the reply boxes and button still stays disabled as if it’s still sending.

Doesn't happen for me. Inputs are cleared and re-enabled. Could you please provide steps to reproduce? Maybe I'm missing something.

**Comment by lcalaresu** [comment]
I'm still using that branch on my production server.
I sometimes encounter minor problems, but I'm not using the very last version anyway... Will update today.
Is there any know problems to ignore?

**Comment by ChristophWurst** [comment]
> I'm still using that branch on my production server.
> I sometimes encounter minor problems, but I'm not using the very last version anyway... Will update today.
> Is there any know problems to ignore?

The only problem to ignore is that the background sync does not work, but will be fixed by #413. Other than that everything *should* work :)

**Comment by ChristophWurst** [comment]
> When sending an email, I get the »Message sent« notification but the reply boxes and button still stays disabled as if it’s still sending.

Fixed :)

**Comment by ChristophWurst** [comment]
I think I was able to debug why tests fail here: https://github.com/nextcloud/mail/pull/445

**Comment by ChristophWurst** [comment]
> I think I was able to debug why tests fail here: #445

Voilà: ``All checks have passed``

**Comment by lcalaresu** [comment]
> Voilà: All checks have passed

Yay! that's great!! 
Coverage also starts to looks interesting :))

It starts to look great. No big issues for the last few days except for known issues above... 
Maybe once on my phone but wasn't able to reproduce it so... :thinking: 

Anyway~~ Should update to a more recent version.
Is there any packaged version of the latest version? Coz I can't generate the product right now :/

**Comment by ChristophWurst** [comment]
> Is there any packaged version of the latest version?

[mail.tar.gz](https://github.com/nextcloud/mail/files/1238377/mail.tar.gz)

**Comment by ChristophWurst** [comment]
> When using a gmail account, I cannot open the [Gmail] folder

Solved meanwhile (can't remember whether it happened in this PR or another one) and now those special folders cannot be opened anymore. However, it's still possible to expand it to open subfolders.

> On a personal server, I have a JS error every time I change the folder (even base folders like sent, inbox, etc.):

Please check whether this is still the case, it could have been fixed meanwhile.

Are those the only two issues you have?

**Comment by ChristophWurst** [comment]
> No big issues for the last few days except for known issues above...

If it's just minor things that I'd say we can fix them later. IMO this PR is open for quite long time and it's about time to finish this ;-)

**Comment by lcalaresu** [comment]
> If it's just minor things that I'd say we can fix them later. IMO this PR is open for quite long time and it's about time to finish this ;-)

I do agree with that one ^_^
Since I installed the version of the mail app from yesterday, I haven't had any probleme yet ^_^

**Comment by ChristophWurst** [comment]
> I do agree with that one ^^
> Since I installed the version of the mail app from yesterday, I haven't had any probleme yet ^^

Yay, thanks a lot for testing. I'll merge this PR now. @jancborchardt please file issues for the issues you had if you can still reproduce them. Thanks everyone :)

**Pull request merged**